contractcourt/anchor_resolver.go

```diff
@@ -0,0 +1,159 @@
+	chanPoint wire.OutPoint
+
+	contractResolverKit
+}
+
+// newAnchorResolver instantiates a new anchor resolver.
+func newAnchorResolver(anchorSignDescriptor input.SignDescriptor,
+	anchor wire.OutPoint, broadcastHeight uint32,
+	chanPoint wire.OutPoint, resCfg ResolverConfig) *anchorResolver {
+
+	r := &anchorResolver{
+		contractResolverKit:  *newContractResolverKit(resCfg),
+		anchorSignDescriptor: anchorSignDescriptor,
+		anchor:               anchor,
+		broadcastHeight:      broadcastHeight,
+		chanPoint:            chanPoint,
+	}
+
+	r.initLogger(r)
+
+	return r
+}
+
+// ResolverKey returns an identifier which should be globally unique for this
+// particular resolver within the chain the original contract resides within.
+func (c *anchorResolver) ResolverKey() []byte {
+	// The anchor resolver is stateless and doesn't need a database key.
+
+	return nil
+}
+
+// Resolve offers the anchor output to the sweeper and waits for it to be swept.
+//
+// NOTE: This function MUST be run as a goroutine.
+func (c *anchorResolver) Resolve() (ContractResolver, error) {
+	// Attempt to update the sweep parameters to the post-confirmation
+	// situation. We don't want to force sweep anymore, because the anchor
+	// lost its special purpose to get the commitment confirmed. It is just
+	// an output that we want to sweep only if it is economical to do so. An
+	// exclusive group is not necessary either, because we know that this is
+	// the only anchor that can be swept.
+	relayFeeRate := c.Sweeper.RelayFeePerKW()
+
+	postConfSweepParams := sweep.Params{
+		Fee: sweep.FeePreference{
+			FeeRate: relayFeeRate,
+		},
+	}
+
+	resultChan, err := c.Sweeper.UpdateParams(
+		c.anchor, postConfSweepParams,
+	)
+
+	// After a restart or when the remote force closes, the sweeper is not
+	// yet aware of the anchor. In that case, offer it as a new input to the
+	// sweeper.
+	if err == lnwallet.ErrNotMine {
+		anchorInput := input.MakeBaseInput(
+			&c.anchor,
+			input.CommitmentAnchor,
+			&c.anchorSignDescriptor,
+			c.broadcastHeight,
+		)
+
+		resultChan, err = c.Sweeper.SweepInput(
+			&anchorInput, postConfSweepParams,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	select {
+	case sweepRes := <-resultChan:
+		switch sweepRes.Err {
+
+		// Anchor was swept successfully.
+		case nil:
+			c.log.Debugf("anchor swept by tx %v",
+				sweepRes.Tx.TxHash())
+
+		// Anchor was swept by someone else. This is possible after the
+		// 16 block csv lock.
+		case sweep.ErrRemoteSpend:
+			c.log.Warnf("our anchor spent by someone else")
+
+		// The sweeper gave up on sweeping the anchor. We consider the
+		// anchor as being lost.
+		case sweep.ErrTooManyAttempts:
+			c.log.Warnf("anchor sweep abandoned")
+
+		// An unexpected error occurred.
+		default:
+			c.log.Errorf("unable to sweep anchor: %v", sweepRes.Err)
+
+			return nil, sweepRes.Err
+		}
+
+	case <-c.quit:
+		return nil, errResolverShuttingDown
+	}
+
+	c.resolved = true
```

Comment:
How will this be communicated on the RPC in the case where we give up? Will the fact that we didn't sweep the anchor show somewhere in the close summary? (explicitly or implicitly )

Reply:
It is shown in the pending close state. The anchor value is part of the limbo or recovered balance and there is also the `anchor_recovered` boolean. However, once the channel is fully resolved that information is no longer accessible. But that is also the case for the outcome of other resolvers.

Reply:
For accounting purposes we should probably add a way of finding this information after it has been resolved (non blocking). I don't know if other resolvers are expected to be resolved in cases where we are unable to sweep them?

Reply:
Yes, @wpaulino had a case with a remote spend in which we are also forced to mark as resolved. Otherwise the channel would stay pending forever.

I consider fixing accounting to be out of scope. It is a whole another project, even just for the channel closes.

Reply:
one day @carlaKC will fix accounting and the world will have `lnd v1.0`